module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +217,111 @@ func TestInvalidateEntity(t *testing.T) {
 	}
 }
 
+// TestAddAndRemoveEntities checks health of heroPool for scenario where entitites are stored and removed in a predetermined order.
+// LRUEjection, NoEjection and RandomEjection are tested. RandomEjection doesn't allow to provide a final state of the pool to check.
+func TestAddAndRemoveEntities(t *testing.T) {
+	for _, tc := range []struct {
+		limit               uint32       // capacity of the pool
+		entityCount         uint32       // total entities to be stored
+		ejectionMode        EjectionMode // ejection mode
+		numberOfOperations  int
+		probabilityOfAdding float32
+	}{
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        LRUEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        NoEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+		{
+			limit:               500,
+			entityCount:         1000,
+			ejectionMode:        RandomEjection,
+			numberOfOperations:  1000,
+			probabilityOfAdding: 0.8,
+		},
+	} {
+		t.Run(fmt.Sprintf("%d-limit-%d-entities", tc.limit, tc.entityCount), func(t *testing.T) {
+			testAddRemoveEntities(t, tc.limit, tc.entityCount, tc.ejectionMode, tc.numberOfOperations, tc.probabilityOfAdding)
+		})
+	}
+}
+
+// Indexes contains an index in the pool of an inserted entity and an index of this entity
+// in the array of entities created by the EntityListFixture.
+type Indexes struct {
+	indexInPool        EIndex
+	indexInEntitiesArr int
+}
+
+func testAddRemoveEntities(t *testing.T, limit uint32, entityCount uint32, ejectionMode EjectionMode, numberOfOperations int, probabilityOfAdding float32) {
+	// create and log the seed.
+	testSeed := time.Now().UnixNano()
+	rand.Seed(testSeed)
+	fmt.Println("Seed used for this testAddRemoveEntities execution", testSeed)
+
+	pool := NewHeroPool(limit, ejectionMode)
+
+	entities := unittest.EntityListFixture(uint(entityCount))
+
+	// an array of random owner Ids.
+	ownerIds := make([]uint64, entityCount, entityCount)
+	// generate ownerId to index in the entities arrray.
+	for i := 0; i < int(entityCount); i++ {
+		ownerIds[i] = rand.Uint64()
+	}
+	// this map maintains entities currently stored in the pool.
+	var addedEntities map[flow.Identifier]Indexes
+	addedEntities = make(map[flow.Identifier]Indexes)
+
+	for i := 0; i < numberOfOperations; i++ {
+		// choose between Add and Remove with a probability of probabilityOfAdding and 1-probabilityOfAdding respectively.
+		if rand.Float32() < probabilityOfAdding || len(addedEntities) == 0 {
+			// adding an entity
+			entityToAdd := rand.Intn(int(entityCount))
+			// check that entity is not already inserted.
+			_, found := addedEntities[entities[entityToAdd].ID()]
+			if !found {
```

Comment:
Test should not skip adding an entity if it has already been added. Instead, it must try picking another entity, and keep trying for a while till either it finds an entity that has not been added to the cache, or exhausts its retries, and fails. Example: 

```go
	entityToAddIndex := -1
	for retryTime := 0; retryTime < 100; retryTime++ {
		entityToAdd := rand.Intn(int(entityCount))
		_, found := addedEntities[entities[entityToAdd].ID()]
		if !found {
			continue
		}
		entityToAddIndex = entityToAdd
	}
	require.GreaterOrEqual(t, entityToAddIndex, 0, "could not find an entity to add after retrying 100 times")
```

Reply:
I am not sure that there'll be any benefit. The test is random anyway. I'll implement your suggestion but without 
```require.GreaterOrEqual(t, entityToAddIndex, 0, "could not find an entity to add after retrying 100 times")``` if number of entities is less than the pool size this condition will fail. Please let me know if you agree with this resolution.

Reply:
Refactored in my proposed changes: https://github.com/onflow/flow-go/pull/4470#discussion_r1263014762

Reply:
Test is failing now for ```limit: 1000, entityCount: 500```